src/lua/LuaFileSystem.cpp

```diff
@@ -60,49 +134,37 @@ static void push_date_time(lua_State *l, const Time::DateTime &dt)
  */
 static int l_filesystem_read_dir(lua_State *l)
 {
-	LuaFileSystem::Root root = static_cast<LuaFileSystem::Root>(LuaConstants::GetConstantFromArg(l, "FileSystemRoot", 1));
-	std::string path;
-	if (lua_gettop(l) > 1)
-		path = luaL_checkstring(l, 2);
-
-	FileSystem::FileSource *fs = nullptr;
-	switch (root) {
-	case LuaFileSystem::ROOT_USER:
-		fs = &FileSystem::userFiles;
-		break;
-
-	case LuaFileSystem::ROOT_DATA:
-		fs = &FileSystem::gameDataFiles;
-		break;
-
-	default:
-		assert(0); // can't happen
-		return 0;
+	SplitURI split_uri = ParseURIArg(l, 1);
+	const FileSource fs = DetermineFileSource(split_uri.scheme);
+	if (nullptr == fs.source)
+	{
+		return luaL_error(l, "Path '%s' does not contain a valid scheme, must be user:// or data://", split_uri.full_path.data());
 	}
 
-	assert(fs);
+	// shame these methods can't take a string_view.
+	std::string rel_path = std::string(split_uri.rel_path);
 
 	{
 		try {
-			const FileSystem::FileInfo &info = fs->Lookup(path);
+			FileSystem::FileInfo info = fs.source->Lookup(rel_path);
 			if (!info.IsDir()) {
-				luaL_error(l, "'%s' is not a directory", path.c_str());
-				return 0;
+				return luaL_error(l, "'%s' is not a directory", split_uri.full_path.data());
 			}
 		} catch (const std::invalid_argument &) {
-			luaL_error(l, "'%s' is not a valid path", path.c_str());
-			return 0;
+			return luaL_error(l, "'%s' is not a valid path", split_uri.full_path.data());
 		}
 	}
 
-	FileSystem::FileEnumerator files(*fs, FileSystem::FileEnumerator::IncludeDirs);
-	files.AddSearchRoot(path);
+	FileSystem::FileEnumerator files(*fs.source, FileSystem::FileEnumerator::IncludeDirs);
+	files.AddSearchRoot(rel_path);
```

Comment:
This can also be expressed as:

```suggestion
auto files = fs.source->Enumerate(rel_path, FileSystem::FileEnumerator::IncludeDirs);
```

The returned `FileEnumerator` object can also be used with a range-for expression either by name or directly, e.g.:

```c++
for (const auto &info : fs.source->Enumerate(rel_path, FileSystem::FileEnumerator::IncludeDirs)) {
    // ...
}
```

This was implemented fairly recently, so you won't see much code using that pattern (yet) but it's one I'd prefer to move to.
(This uses a temporary adapter object with a reference to the temporary enumerator object, which AFAIK according to temporary lifetime rules will not be destroyed until the end of the range-for statement.)

Reply:
Nice, I hadn't really touched this but I've now moved it to the new idiom